### PR TITLE
task instance listener: call on_task_instance_running after rendering templates

### PR DESCRIPTION
**Pull request by mobuchowski**
Previously, `on_task_instance_running` was being called in `_run_raw_task`. After this, it will be running close to where `pre_execute` callback is called.

This had disadvantage of being run before operator templates are rendered. 
In some usages, this required users of this interface (like OpenLineage) to resolve templates themselves, which comes with obvious performance drawback. Now, the state of actual executed task is the same as on `execute`. 

Another reason of moving call site is that idea of task instance listener API is to provide notification as close to when actual execution happens. This is as close as we can get AFAIK.

### CODE REVIEW

File: airflow/models/taskinstance.py [hunk 1]
```diff
@@ -1601,7 +1596,8 @@ def _register_dataset_changes(self, *, session: Session) -> None:
                     session=session,
                 )
 
-    def _execute_task_with_callbacks(self, context, test_mode=False):
+    @provide_session
+    def _execute_task_with_callbacks(self, context, test_mode: bool = False, session: Session = NEW_SESSION):
```
Comment (uranusjr) [inline airflow/models/taskinstance.py]:
```suggestion
    def _execute_task_with_callbacks(self, context, test_mode: bool = False, *, session: Session):
```

This is only ever called with an explicit session, no point decorating.

File: airflow/models/taskinstance.py [hunk 2]
```diff
@@ -1498,14 +1498,9 @@ def _run_raw_task(
...
 
-        # We lose previous state because it's changed in other process in LocalTaskJob.
-        # We could probably pass it through here though...
-        get_listener_manager().hook.on_task_instance_running(
-            previous_state=TaskInstanceState.QUEUED, task_instance=self, session=session
-        )
         try:
             if not mark_success:
-                self._execute_task_with_callbacks(context, test_mode)
+                self._execute_task_with_callbacks(context, test_mode, session)
```
Comment (uranusjr) [inline airflow/models/taskinstance.py]:
```suggestion
                self._execute_task_with_callbacks(context, test_mode, session=session)
```

**Review by uranusjr — DISMISSED** [review]

File: airflow/models/taskinstance.py (hunk 1 above)
Comment (mobuchowski) [reply airflow/models/taskinstance.py]: Fixed.

File: airflow/models/taskinstance.py (hunk 2 above)
Comment (mobuchowski) [reply airflow/models/taskinstance.py]: Fixed.

File: airflow/models/taskinstance.py [hunk 3]
```diff
@@ -1651,16 +1646,23 @@ def signal_handler(signum, frame):
                 )
 
             # Run pre_execute callback
-            self.task.pre_execute(context=context)
+            # Is never MappedOperator at this point
+            self.task.pre_execute(context=context)  # type: ignore[union-attr]
```
Comment (uranusjr) [inline airflow/models/taskinstance.py:1650]: Where is the task unmapped before this line is reached? How can we be sure this is the case?

Comment (mobuchowski) [reply airflow/models/taskinstance.py:1650]: @uranusjr I'm actually going by your code 😉  - render templates is just before that.

https://github.com/apache/airflow/commit/5560a46bfe8a14205c5e8a14f0b5c2ae74ee100c

Comment (uranusjr) [reply airflow/models/taskinstance.py:1650]: The fact I wrote the code is not a particularly promising guarantee though 😛 

But code here is pretty awful. I’ll add a commit to better fix this up later…

**Review by uranusjr — APPROVED** [review]

**Pull request merged**